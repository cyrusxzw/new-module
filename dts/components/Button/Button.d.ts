import React from 'react';
import type { ButtonProps } from './Button.types';
declare const Button: React.ForwardRefExoticComponent<
  ButtonProps & React.RefAttributes<HTMLButtonElement>
>;
export { Button };
