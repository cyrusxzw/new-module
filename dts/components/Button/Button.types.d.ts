import { ReactNode, MouseEvent } from 'react';
import type { Themes } from '~/types';
declare type ButtonBehaviourType = 'button' | 'reset' | 'submit';
declare type ButtonProps = {
  children?: ReactNode;
  className?: string;
  dataTestRef?: string;
  id?: string;
  isAlternate?: boolean;
  isEnabled?: boolean;
  isInline?: boolean;
  onClick?: (event: MouseEvent) => void;
  tabIndex?: number;
  title?: string;
  type?: ButtonBehaviourType;
  theme?: Themes;
};
export type { ButtonProps };
