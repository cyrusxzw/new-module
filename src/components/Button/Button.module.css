@import 'styles/index';

.base {
  position: relative;
  padding: 0;
  border: 0;
  appearance: none;
  background-color: transparent;
  cursor: pointer;
  font-family: var(--font-family-sans-serif-medium);
  font-size: rem(14px);
  font-weight: var(--font-weight-normal);
  line-height: 1.4;
  outline: none;
  text-decoration: none;

  > * {
    pointer-events: none;
  }

  &,
  & * {
    box-sizing: border-box;
  }
}

.blockStyle {
  display: inline-block;
  min-width: 210px;
  max-width: 300px;
  height: 60px;
  padding: 19px 23px;
  border-width: 1px;
  border-style: solid;
  transition: all 200ms var(--easing-ease-in-out-cubic);

  @media (--viewport-xs-sm-only) {
    width: 100%;
    max-width: 100%;
  }
}

.dark {
  @extend %font-theme-dark-link;

  &.blockStyle {
    border-color: var(--color-grey-rgba-20);

    &:hover,
    &:focus {
      background-color: var(--color-dark-background);
      color: var(--color-light-link-hover);
    }

    &.alternate {
      border-color: var(--color-dark-background);
      background-color: var(--color-dark-background);
      color: var(--color-light-link-hover);

      &:hover,
      &:focus {
        border-color: var(--color-black);
        background-color: var(--color-black);
      }
    }
  }
}

.light {
  @extend %font-theme-light-link;

  &.blockStyle {
    border-color: var(--color-white);

    &:focus,
    &:hover {
      background-color: var(--color-white);
      color: var(--color-dark-copy);
    }
  }

  &.alternate {
    border-color: var(--color-light-link-hover);
    background-color: var(--color-light-link-hover);
    color: var(--color-dark-copy);

    &:hover,
    &:focus {
      border-color: var(--color-white);
      background-color: var(--color-white);
    }
  }
}

.disabled[disabled] {
  background-color: transparent;
  color: var(--color-grey-30);
  cursor: default;

  &.blockStyle {
    border-color: var(--color-grey-rgba-20);

    &.alternate {
      background-color: var(--color-dark-disabled);
      color: var(--color-white);
    }
  }

  &:hover,
  &:focus {
    background-color: transparent;
    color: var(--color-grey-30);

    &.blockStyle {
      border-color: var(--color-grey-rgba-20);

      &.alternate {
        border-color: var(--color-grey-rgba-20);
        background-color: var(--color-dark-disabled);
      }
    }
  }
}
