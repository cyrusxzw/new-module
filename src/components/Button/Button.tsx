import React, { forwardRef } from 'react';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import type { ButtonProps } from './Button.types';
import styles from './Button.module.css';

const Button = forwardRef<HTMLButtonElement, ButtonProps>(function ButtonRef(
  {
    aria,
    children,
    className,
    dataTestRef,
    id,
    isAlternate = false,
    isEnabled = true,
    isInline = false,
    onClick,
    onMouseEnter,
    tabIndex,
    title,
    type = 'button',
    theme,
    ...otherProps
  },
  ref,
) {
  const currentTheme = useThemeContext(theme, 'dark');

  if (!children) return null;

  const classSet = cx(
    styles.base,
    { [styles.alternate]: isAlternate },
    { [styles.blockStyle]: !isInline },
    styles[currentTheme],
    { [styles.disabled]: !isEnabled },
    className,
  );

  return (
    <button
      aria-expanded={aria?.expanded ?? null}
      aria-haspopup={aria?.haspopup ?? null}
      aria-hidden={aria?.hidden ?? null}
      aria-label={aria?.label ?? null}
      className={classSet}
      data-test-ref={dataTestRef}
      disabled={!isEnabled}
      id={id}
      onClick={onClick}
      onMouseEnter={onMouseEnter}
      ref={ref}
      tabIndex={tabIndex}
      title={title}
      type={type}
      {...otherProps}
    >
      {children}
    </button>
  );
});

export { Button };
