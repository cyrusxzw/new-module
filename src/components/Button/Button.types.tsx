import { ReactNode, MouseEvent } from 'react';
import type { Aria, Themes } from '~/types';

type ButtonBehaviourType = 'button' | 'reset' | 'submit';

type ButtonProps = {
  aria?: Pick<Aria, 'expanded' | 'haspopup' | 'hidden' | 'label'>;
  children?: ReactNode;
  className?: string;
  dataTestRef?: string;
  id?: string;
  isAlternate?: boolean;
  isEnabled?: boolean;
  isInline?: boolean;
  onClick?: (event: MouseEvent) => void;
  tabIndex?: number;
  title?: string;
  type?: ButtonBehaviourType;
  theme?: Themes;
};

export type { ButtonProps };
