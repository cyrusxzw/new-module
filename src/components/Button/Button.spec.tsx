import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Button } from './Button';

describe('<Button />', () => {
  const handleOnClick = jest.fn();
  const buttonText = 'button';

  it('should render base component correctly with `onClick` prop', () => {
    const { container } = render(
      <Button className="class" onClick={handleOnClick} title="button-title">
        {buttonText}
      </Button>,
    );

    const button = screen.getByRole('button', { name: /button/i });

    expect(container).toMatchSnapshot();
    expect(button).toBeTruthy();
  });

  it('should return `null` if no children are passed', () => {
    render(
      <Button className="aesop" onClick={handleOnClick} title="button-title" />,
    );

    const button = screen.queryByRole('button', { name: /button/i });

    expect(button).not.toBeInTheDocument();
  });

  it('should call handle onClick callback if pressed', () => {
    render(
      <Button className="class" onClick={handleOnClick} title="button-title">
        {buttonText}
      </Button>,
    );

    const button = screen.getByRole('button', { name: /button/i });

    userEvent.click(button);

    expect(handleOnClick).toHaveBeenCalledTimes(1);
  });
});
