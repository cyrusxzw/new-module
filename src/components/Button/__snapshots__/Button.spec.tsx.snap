// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render base component correctly with \`onClick\` prop 1`] = `
<div>
  <button
    class="base blockStyle dark class"
    title="button-title"
    type="button"
  >
    button
  </button>
</div>
`;
